File: src/video_editor.py
from .video_io import VideoIO
from .geometric_ops import GeometricOps
from .temporal_ops import TemporalOps
from .color_ops import ColorOps
from .filter_ops import FilterOps
from .morph_ops import MorphOps
from .composite_ops import CompositeOps

class VideoEditor:
    def __init__(self, input_path):
        self.io = VideoIO(input_path)
        self.geom = GeometricOps()
        self.temp = TemporalOps()
        self.color = ColorOps()
        self.filt = FilterOps()
        self.morph = MorphOps()
        self.comp = CompositeOps()
    
    def execute(self, cmd, output, args):
        if cmd == 'crop':
            frames = self.io.read_all()
            frames = self.geom.crop(frames, int(args[0]), int(args[1]), int(args[2]), int(args[3]))
            self.io.write(output, frames)
        
        elif cmd == 'resize':
            frames = self.io.read_all()
            frames = self.geom.resize(frames, int(args[0]), int(args[1]))
            self.io.write(output, frames)
        
        elif cmd == 'rotate':
            frames = self.io.read_all()
            frames = self.geom.rotate(frames, float(args[0]), self.io.w, self.io.h)
            self.io.write(output, frames)
        
        elif cmd == 'flip':
            frames = self.io.read_all()
            d = int(args[0]) if args else 1
            frames = self.geom.flip(frames, d)
            self.io.write(output, frames)
        
        elif cmd == 'reverse':
            frames = self.io.read_all()
            frames = self.temp.reverse(frames)
            self.io.write(output, frames)
        
        elif cmd == 'speed':
            frames = self.io.read_all()
            s = float(args[0])
            if s >= 1:
                frames = self.temp.speed_up(frames, s)
                self.io.write(output, frames)
            else:
                self.io.write_fps(output, frames, int(self.io.fps * s))
        
        elif cmd == 'trim':
            frames = self.io.read_range(int(args[0]), int(args[1]))
            self.io.write(output, frames)
        
        elif cmd == 'gray':
            frames = self.io.read_all()
            frames = self.color.gray(frames)
            self.io.write(output, frames)
        
        elif cmd == 'invert':
            frames = self.io.read_all()
            frames = self.color.invert(frames)
            self.io.write(output, frames)
        
        elif cmd == 'bright':
            frames = self.io.read_all()
            v = int(args[0]) if args else 50
            frames = self.color.brightness(frames, v)
            self.io.write(output, frames)
        
        elif cmd == 'contrast':
            frames = self.io.read_all()
            v = float(args[0]) if args else 1.5
            frames = self.color.contrast(frames, v)
            self.io.write(output, frames)
        
        elif cmd == 'hue':
            frames = self.io.read_all()
            frames = self.color.hue_shift(frames, int(args[0]))
            self.io.write(output, frames)
        
        elif cmd == 'sat':
            frames = self.io.read_all()
            frames = self.color.saturation(frames, float(args[0]))
            self.io.write(output, frames)
        
        elif cmd == 'blur':
            frames = self.io.read_all()
            k = int(args[0]) if args else 15
            frames = self.filt.blur(frames, k)
            self.io.write(output, frames)
        
        elif cmd == 'sharpen':
            frames = self.io.read_all()
            frames = self.filt.sharpen(frames)
            self.io.write(output, frames)
        
        elif cmd == 'edge':
            frames = self.io.read_all()
            frames = self.filt.edge(frames)
            self.io.write(output, frames)
        
        elif cmd == 'denoise':
            frames = self.io.read_all()
            frames = self.filt.denoise(frames)
            self.io.write(output, frames)
        
        elif cmd == 'threshold':
            frames = self.io.read_all()
            t = int(args[0]) if args else 127
            frames = self.filt.threshold(frames, t)
            self.io.write(output, frames)
        
        elif cmd == 'dilate':
            frames = self.io.read_all()
            frames = self.morph.dilate(frames, int(args[0]))
            self.io.write(output, frames)
        
        elif cmd == 'erode':
            frames = self.io.read_all()
            frames = self.morph.erode(frames, int(args[0]))
            self.io.write(output, frames)
        
        elif cmd == 'open':
            frames = self.io.read_all()
            frames = self.morph.open(frames, int(args[0]))
            self.io.write(output, frames)
        
        elif cmd == 'close':
            frames = self.io.read_all()
            frames = self.morph.close(frames, int(args[0]))
            self.io.write(output, frames)
        
        elif cmd == 'stack':
            frames1 = self.io.read_all()
            io2 = VideoIO(args[0])
            frames2 = io2.read_all()
            frames = self.comp.stack_horizontal(frames1, frames2)
            self.io.write(output, frames)
        
        elif cmd == 'overlay':
            frames1 = self.io.read_all()
            io2 = VideoIO(args[0])
            frames2 = io2.read_all()
            x = int(args[1]) if len(args) > 1 else 0
            y = int(args[2]) if len(args) > 2 else 0
            frames = self.comp.overlay(frames1, frames2, x, y)
            self.io.write(output, frames)
        
        elif cmd == 'zoom':
            frames = self.io.read_all()
            factor = float(args[0]) if args else 1.5
            cx = int(args[1]) if len(args) > 1 else None
            cy = int(args[2]) if len(args) > 2 else None
            frames = self.geom.zoom(frames, factor, cx, cy)
            self.io.write(output, frames)
        
        elif cmd == 'pan':
            frames = self.io.read_all()
            dx = int(args[0]) if args else 0
            dy = int(args[1]) if len(args) > 1 else 0
            frames = self.geom.pan(frames, dx, dy)
            self.io.write(output, frames)
        
        elif cmd == 'freeze':
            frames = self.io.read_all()
            frame_idx = int(args[0]) if args else 0
            duration = int(args[1]) if len(args) > 1 else 30
            frames = self.temp.freeze(frames, frame_idx, duration)
            self.io.write(output, frames)
        
        elif cmd == 'loop':
            frames = self.io.read_all()
            count = int(args[0]) if args else 2
            frames = self.temp.loop(frames, count)
            self.io.write(output, frames)
        
        elif cmd == 'duplicate':
            frames = self.io.read_all()
            factor = int(args[0]) if args else 2
            frames = self.temp.duplicate(frames, factor)
            self.io.write(output, frames)
        
        elif cmd == 'sepia':
            frames = self.io.read_all()
            frames = self.color.sepia(frames)
            self.io.write(output, frames)
        
        elif cmd == 'warm':
            frames = self.io.read_all()
            intensity = int(args[0]) if args else 20
            frames = self.color.warm(frames, intensity)
            self.io.write(output, frames)
        
        elif cmd == 'cool':
            frames = self.io.read_all()
            intensity = int(args[0]) if args else 20
            frames = self.color.cool(frames, intensity)
            self.io.write(output, frames)
        
        elif cmd == 'posterize':
            frames = self.io.read_all()
            levels = int(args[0]) if args else 4
            frames = self.color.posterize(frames, levels)
            self.io.write(output, frames)
        
        elif cmd == 'emboss':
            frames = self.io.read_all()
            frames = self.filt.emboss(frames)
            self.io.write(output, frames)
        
        elif cmd == 'cartoon':
            frames = self.io.read_all()
            frames = self.filt.cartoon(frames)
            self.io.write(output, frames)
        
        elif cmd == 'vignette':
            frames = self.io.read_all()
            intensity = float(args[0]) if args else 0.3
            frames = self.filt.vignette(frames, intensity)
            self.io.write(output, frames)
        
        elif cmd == 'motion_blur':
            frames = self.io.read_all()
            size = int(args[0]) if args else 15
            angle = float(args[1]) if len(args) > 1 else 0
            frames = self.filt.motion_blur(frames, size, angle)
            self.io.write(output, frames)
        
        elif cmd == 'fade_in':
            frames = self.io.read_all()
            duration = int(args[0]) if args else 30
            frames = self.comp.fade_in(frames, duration)
            self.io.write(output, frames)
        
        elif cmd == 'fade_out':
            frames = self.io.read_all()
            duration = int(args[0]) if args else 30
            frames = self.comp.fade_out(frames, duration)
            self.io.write(output, frames)
        
        elif cmd == 'crossfade':
            frames1 = self.io.read_all()
            io2 = VideoIO(args[0])
            frames2 = io2.read_all()
            duration = int(args[1]) if len(args) > 1 else 30
            frames = self.comp.crossfade(frames1, frames2, duration)
            self.io.write(output, frames)
        
        elif cmd == 'stack_v':
            frames1 = self.io.read_all()
            io2 = VideoIO(args[0])
            frames2 = io2.read_all()
            frames = self.comp.stack_vertical(frames1, frames2)
            self.io.write(output, frames)
        
        elif cmd == 'text':
            frames = self.io.read_all()
            text = args[0] if args else "Text"
            x = int(args[1]) if len(args) > 1 else 10
            y = int(args[2]) if len(args) > 2 else 30
            scale = float(args[3]) if len(args) > 3 else 1.0
            color = (int(args[4]), int(args[5]), int(args[6])) if len(args) > 6 else (255, 255, 255)
            thickness = int(args[7]) if len(args) > 7 else 2
            frames = self.comp.text_overlay(frames, text, x, y, scale, color, thickness)
            self.io.write(output, frames)
        
        elif cmd == 'extract':
            frame_idx = int(args[0]) if args else 0
            self.io.extract_frame(frame_idx, output)
            print(f"Done: {output}")

File: src/VideoEditor.cpp
#include "../include/VideoEditor_structured.h"
#include "VideoIO.h"
#include "GeometricOps.h"
#include "TemporalOps.h"
#include "ColorOps.h"
#include "FilterOps.h"
#include "MorphOps.h"
#include "CompositeOps.h"

#include <iostream>

using namespace std;

// Impl is implemented here and uses the free classes above
class VideoEditor::Impl {
public:
    VideoIO* io;
    Impl(const string& inputPath) { io = new VideoIO(inputPath); }
    ~Impl() { delete io; }

    void execute(const string& cmd, const string& output, const vector<string>& args) {
        if (cmd == "crop") {
            auto frames = io->readAll();
            frames = GeometricOps::crop(frames, stoi(args[0]), stoi(args[1]), stoi(args[2]), stoi(args[3]));
            io->write(output, frames);
        }
        else if (cmd == "resize") {
            auto frames = io->readAll();
            frames = GeometricOps::resize(frames, stoi(args[0]), stoi(args[1]));
            io->write(output, frames);
        }
        else if (cmd == "rotate") {
            auto frames = io->readAll();
            frames = GeometricOps::rotate(frames, stod(args[0]), io->w, io->h);
            io->write(output, frames);
        }
        else if (cmd == "flip") {
            auto frames = io->readAll();
            int d = args.empty() ? 1 : stoi(args[0]);
            frames = GeometricOps::flip(frames, d);
            io->write(output, frames);
        }
        else if (cmd == "reverse") {
            auto frames = io->readAll();
            frames = TemporalOps::reverse(frames);
            io->write(output, frames);
        }
        else if (cmd == "speed") {
            auto frames = io->readAll();
            double s = stod(args[0]);
            if (s >= 1) { frames = TemporalOps::speedUp(frames, s); io->write(output, frames); }
            else io->writeFps(output, frames, (int)(io->fps * s));
        }
        else if (cmd == "trim") { auto frames = io->readRange(stoi(args[0]), stoi(args[1])); io->write(output, frames); }
        else if (cmd == "gray") { auto frames = io->readAll(); frames = ColorOps::gray(frames); io->write(output, frames); }
        else if (cmd == "invert") { auto frames = io->readAll(); frames = ColorOps::invert(frames); io->write(output, frames); }
        else if (cmd == "bright") { auto frames = io->readAll(); int v = args.empty() ? 50 : stoi(args[0]); frames = ColorOps::brightness(frames, v); io->write(output, frames); }
        else if (cmd == "contrast") { auto frames = io->readAll(); double v = args.empty() ? 1.5 : stod(args[0]); frames = ColorOps::contrast(frames, v); io->write(output, frames); }
        else if (cmd == "hue") { auto frames = io->readAll(); frames = ColorOps::hueShift(frames, stoi(args[0])); io->write(output, frames); }
        else if (cmd == "sat") { auto frames = io->readAll(); frames = ColorOps::saturation(frames, stod(args[0])); io->write(output, frames); }
        else if (cmd == "blur") { auto frames = io->readAll(); int k = args.empty() ? 15 : stoi(args[0]); frames = FilterOps::blur(frames, k); io->write(output, frames); }
        else if (cmd == "sharpen") { auto frames = io->readAll(); frames = FilterOps::sharpen(frames); io->write(output, frames); }
        else if (cmd == "edge") { auto frames = io->readAll(); frames = FilterOps::edge(frames); io->write(output, frames); }
        else if (cmd == "denoise") { auto frames = io->readAll(); frames = FilterOps::denoise(frames); io->write(output, frames); }
        else if (cmd == "threshold") { auto frames = io->readAll(); int t = args.empty() ? 127 : stoi(args[0]); frames = FilterOps::threshold(frames, t); io->write(output, frames); }
        else if (cmd == "dilate") { auto frames = io->readAll(); frames = MorphOps::dilate(frames, stoi(args[0])); io->write(output, frames); }
        else if (cmd == "erode") { auto frames = io->readAll(); frames = MorphOps::erode(frames, stoi(args[0])); io->write(output, frames); }
        else if (cmd == "open") { auto frames = io->readAll(); frames = MorphOps::open(frames, stoi(args[0])); io->write(output, frames); }
        else if (cmd == "close") { auto frames = io->readAll(); frames = MorphOps::close(frames, stoi(args[0])); io->write(output, frames); }
        else if (cmd == "stack") { auto frames1 = io->readAll(); VideoIO io2(args[0]); auto frames2 = io2.readAll(); auto outFrames = CompositeOps::stackHorizontal(frames1, frames2); io->write(output, outFrames); }
        else if (cmd == "overlay") { auto frames1 = io->readAll(); VideoIO io2(args[0]); auto frames2 = io2.readAll(); int x = args.size() > 1 ? stoi(args[1]) : 0; int y = args.size() > 2 ? stoi(args[2]) : 0; auto outFrames = CompositeOps::overlay(frames1, frames2, x, y); io->write(output, outFrames); }
        else if (cmd == "zoom") { auto frames = io->readAll(); double factor = args.empty() ? 1.5 : stod(args[0]); int cx = args.size() > 1 ? stoi(args[1]) : -1; int cy = args.size() > 2 ? stoi(args[2]) : -1; frames = GeometricOps::zoom(frames, factor, cx, cy); io->write(output, frames); }
        else if (cmd == "pan") { auto frames = io->readAll(); int dx = args.empty() ? 0 : stoi(args[0]); int dy = args.size() > 1 ? stoi(args[1]) : 0; frames = GeometricOps::pan(frames, dx, dy); io->write(output, frames); }
        else if (cmd == "freeze") { auto frames = io->readAll(); int frameIdx = args.empty() ? 0 : stoi(args[0]); int duration = args.size() > 1 ? stoi(args[1]) : 30; frames = TemporalOps::freeze(frames, frameIdx, duration); io->write(output, frames); }
        else if (cmd == "loop") { auto frames = io->readAll(); int count = args.empty() ? 2 : stoi(args[0]); frames = TemporalOps::loop(frames, count); io->write(output, frames); }
        else if (cmd == "duplicate") { auto frames = io->readAll(); int factor = args.empty() ? 2 : stoi(args[0]); frames = TemporalOps::duplicate(frames, factor); io->write(output, frames); }
        else if (cmd == "sepia") { auto frames = io->readAll(); frames = ColorOps::sepia(frames); io->write(output, frames); }
        else if (cmd == "warm") { auto frames = io->readAll(); int intensity = args.empty() ? 20 : stoi(args[0]); frames = ColorOps::warm(frames, intensity); io->write(output, frames); }
        else if (cmd == "cool") { auto frames = io->readAll(); int intensity = args.empty() ? 20 : stoi(args[0]); frames = ColorOps::cool(frames, intensity); io->write(output, frames); }
        else if (cmd == "posterize") { auto frames = io->readAll(); int levels = args.empty() ? 4 : stoi(args[0]); frames = ColorOps::posterize(frames, levels); io->write(output, frames); }
        else if (cmd == "emboss") { auto frames = io->readAll(); frames = FilterOps::emboss(frames); io->write(output, frames); }
        else if (cmd == "cartoon") { auto frames = io->readAll(); frames = FilterOps::cartoon(frames); io->write(output, frames); }
        else if (cmd == "vignette") { auto frames = io->readAll(); double intensity = args.empty() ? 0.3 : stod(args[0]); frames = FilterOps::vignette(frames, intensity); io->write(output, frames); }
        else if (cmd == "motion_blur") { auto frames = io->readAll(); int size = args.empty() ? 15 : stoi(args[0]); double angle = args.size() > 1 ? stod(args[1]) : 0.0; frames = FilterOps::motionBlur(frames, size, angle); io->write(output, frames); }
        else if (cmd == "fade_in") { auto frames = io->readAll(); int duration = args.empty() ? 30 : stoi(args[0]); frames = CompositeOps::fadeIn(frames, duration); io->write(output, frames); }
        else if (cmd == "fade_out") { auto frames = io->readAll(); int duration = args.empty() ? 30 : stoi(args[0]); frames = CompositeOps::fadeOut(frames, duration); io->write(output, frames); }
        else if (cmd == "crossfade") { auto frames1 = io->readAll(); VideoIO io2(args[0]); auto frames2 = io2.readAll(); int duration = args.size() > 1 ? stoi(args[1]) : 30; auto outF = CompositeOps::crossfade(frames1, frames2, duration); io->write(output, outF); }
        else if (cmd == "stack_v") { auto frames1 = io->readAll(); VideoIO io2(args[0]); auto frames2 = io2.readAll(); auto outF = CompositeOps::stackVertical(frames1, frames2); io->write(output, outF); }
        else if (cmd == "text") { auto frames = io->readAll(); string text = args.empty() ? "Text" : args[0]; int x = args.size() > 1 ? stoi(args[1]) : 10; int y = args.size() > 2 ? stoi(args[2]) : 30; double scale = args.size() > 3 ? stod(args[3]) : 1.0; Scalar color = args.size() > 6 ? Scalar(stoi(args[6]), stoi(args[5]), stoi(args[4])) : Scalar(255,255,255); int thickness = args.size() > 7 ? stoi(args[7]) : 2; auto outF = CompositeOps::textOverlay(frames, text, x, y, scale, color, thickness); io->write(output, outF); }
        else if (cmd == "extract") { int frameIdx = args.empty() ? 0 : stoi(args[0]); io->extractFrame(frameIdx, output); cout << "Done: " << output << endl; return; }
        else { cerr << "Unknown command: " << cmd << endl; return; }
        cout << "Done: " << output << endl;
    }
};

VideoEditor::VideoEditor(const std::string& inputPath) { impl = new Impl(inputPath); }
VideoEditor::~VideoEditor() { delete impl; }
void VideoEditor::execute(const std::string& cmd, const std::string& output, const std::vector<std::string>& args) { impl->execute(cmd, output, args); }
